sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1446,15 +1446,15 @@ class Dataset[T] private[sql](
       if (sqlContext.conf.supportQuotedRegexColumnName) {
         colRegex(colName)
       } else {
-        Column(addDataFrameIdToCol(resolve(colName)))
+        Column(addPlanId(UnresolvedAttribute.quotedString(colName)))
       }
   }
 
   /**
    * Selects column based on the column index (0-based) and returns it as a [[Column]].
    */
   private[sql] def col(index: Int): Column = {
-    Column(addDataFrameIdToCol(queryExecution.analyzed.output(index)))
```

Comment:
right now PLAN_ID is not actually used in GetColumnByOrdinal resolution, maybe we should  add it in 
https://github.com/apache/spark/blob/2a10c8d93aa9033842471e4f676fddb3b3f90940/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala#L493-L506

Reply:
where do we use it? I feel it's risky to get dataframe column references by ordinal. If it's ambiguous, there is no way to fix it, as it can't use qualified names.

Reply:
it was introduced in https://github.com/apache/spark/pull/42828, only used in Python side

Reply:
PySpark already provided `df[i]` with a `int i` to get the i-th column;
and PySpark support order by ordinal `df.orderBy(1, 2, 3)`;

`GetColumnByOrdinal` is used to support duplicated column names in above two cases.